quinn-proto/src/connection/mod.rs

```diff
@@ -402,67 +405,166 @@ where
+            } else {
+                false
+            };
 
-        let mut congestion_blocked = false;
+            // Is there data or a close message to send in this space?
+            if !self.space_can_send(space_id) && !close_in_space {
+                space_idx += 1;
+                continue;
+            }
 
-        for space_id in spaces {
             let mut ack_eliciting =
                 !self.spaces[space_id].pending.is_empty() || self.spaces[space_id].ping_pending;
             if space_id == SpaceId::Data {
                 ack_eliciting |= self.can_send_1rtt();
+            }
+
+            // Can we append more data into the current buffer?
+            // It is not safe to assume that `buf.len()` is the end of the data,
+            // since the last packet might not have been finished.
+            let buf_end = if let Some(builder) = &builder {
+                buf.len().max(builder.min_size) + builder.tag_len
+            } else {
+                buf.len()
+            };
+
+            if !coalesce || buf_capacity - buf_end < MIN_PACKET_SPACE {
+                // We need to send 1 more datagram and extend the buffer for that.
+
+                // Is 1 more datagram allowed?
+                if buf_capacity >= self.path.mtu as usize * MAX_DATAGRAMS {
+                    // No more datagrams allowed
+                    break;
+                }
+
+                // Anti-amplification is only based on `total_sent`, which gets
+                // updated at the end of this method. Therefore we pass the accumulated
+                // amount of datagrams and bytes here.
+                if self
+                    .path
+                    .anti_amplification_blocked(self.path.mtu as u64 * (num_datagrams + 1) as u64)
+                {
+                    trace!("blocked by anti-amplification");
+                    break;
+                }
+
+                // Congestion control and pacing checks
                 // Tail loss probes must not be blocked by congestion, or a deadlock could arise
                 if ack_eliciting && self.spaces[space_id].loss_probes == 0 {
-                    if self.congestion_blocked(u64::from(self.path.mtu)) {
+                    // Assume the current packet will get padded to fill the full MTU
+                    let untracked_bytes = if let Some(builder) = &builder {
+                        buf_capacity - builder.partial_encode.start
+                    } else {
+                        0
+                    } as u64;
+                    debug_assert!(untracked_bytes <= self.path.mtu as u64);
+
+                    let bytes_to_send = u64::from(self.path.mtu) + untracked_bytes;
+
+                    if self.congestion_blocked(bytes_to_send) {
+                        space_idx += 1;
                         congestion_blocked = true;
+                        // TODO:Should those rather be break?
+                        // Maybe not due to loss probes
```

Comment:
I think this comment is correct as we might have a loss probe in a higher space, so let's rephrase it as an explanation.

Reply:
makes sense. Done